Implement an atomic input method API

dcz-self opened the pull request:
This is based on the Wayland text-input procotol, where changes are grouped into sets and applied with .commit or .done, depending on the direction.

- [x] Tested on all platforms changed
- [x] Added an entry to the `changelog` module if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality

dcz-self commented:
I updated the inline docs and I couldn't find any separate docs so I checked the box.

I didn't mark the old API as deprecated because I'm not 100% sure how to go about it.

dcz-self commented on winit-core/src/window.rs:
```diff
@@ -1127,6 +1127,23 @@ pub trait Window: AsAny + Send + Sync + fmt::Debug {
+    ///
+    /// Input methods allows the user to compose text without using a keyboard. Requesting one may be beneficial for touch screen environments or ones where, for example, East Asian scripts may be entered.
+    ///
+    /// Switching from one text input area to another in the same window should cause the disabling of the input method and the enabling again in the other context.
```
I think it's unclear, nothing here should trigger on resizing. This note was meant to say: if the focus within the application switches from one logical text field to another, then the input method should be informed of it by disabling and enabling it again. I'll try to submit a clearer description.

dcz-self commented on winit-core/src/event.rs:
```diff
@@ -955,6 +964,70 @@ pub enum Ime {
+///
+/// Another significant usage is breaking the direct relationships between keys and letters, for example, to be able to type thousands of Chinese characters, each with multiple key strokes.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum InputMethodEvent {
+    /// The input method is still enabled or has just become enabled
+    Enabled(InputMethodState),
+    /// The input method has just become disabled
+    Disabled,
+}
```
Do you mean splitting the atomic submission of state to IME and receiving of atomic requests from the IME into separate commits?

kchibisov replied:
I'd say even PRs, because I don't think we need sending state atomically to the user, like users will get calls in order, so it's not that hard for them to figure out the correct thing, atomic serves nothing here, since you basically move the logic on how to apply from winit to the user, while user might not be aware of all the details, and explicit logic like that would require everyone to understand.

When you basically send state to the user like we have now, they won't have issue with processing it in the right order, since we mandate the right order.

dcz-self commented on winit-core/src/window.rs:
```diff
@@ -1528,6 +1545,112 @@ impl Default for ImePurpose {
+            purpose: ImePurpose::Normal,
+            cursor_area: Some(cursor_area),
+        }
+    }
```
Actually no, I'm wrong. I'll change it.

dcz-self commented:
While I wait for the nightly toolchain to download, what should I do with all the alternative platforms? Try to patch in the new functionality without testing? Stub the new methods?
I should put those changes in the same commit as the trait/enum modifications, right?

kchibisov commented on winit-core/src/window.rs:
```diff
@@ -1528,6 +1545,87 @@ impl Default for ImePurpose {
+///
+/// - **iOS / Android / Web / Windows / X11 / macOS / Orbital:** Unsupported.
+#[non_exhaustive]
+#[derive(Debug, PartialEq, Clone, Copy)]
```
Could you drop `Copy`?

kchibisov commented on winit-core/src/window.rs:
```diff
@@ -1528,6 +1545,87 @@ impl Default for ImePurpose {
+        Self { cursor_area: Some((position, size)), ..self }
+    }
+
+    /// Returns current cursor area
+    pub fn cursor_area(&self) -> Option<(Position, Size)> {
+        self.cursor_area
+    }
```
This is redundant.

dcz-self commented:
I removed the bits concerned with receiving events atomically. I think that needs to arrive at some point if features like deleting text are to be supported, though, because otherwise you'll get flicker as changes are updated sequentially.

I also fixed a bug regarding sending the Enabled event and left a little comment.

And I changed the application example to use this new API.

The only thing left to do is to do something about the unsupported platforms.

kchibisov commented on winit-core/src/window.rs:
```diff
@@ -1528,6 +1553,89 @@ impl Default for ImePurpose {
+
+    /// Sets the IME cursor editing area.
+    ///
+    /// On Wayland, this the cursor area should be set from the first request. Otherwise, the input
+    /// method will consider the functionality unsupported.
```
Move it to platform specific.

kchibisov reviewed (changes requested):
The API looks fine, so to proceed with that, let's do the following:

1. Add a deprecated note on old IME methods.
2. Add default trait impls on `dyn Window`calling to `set_ime_state`. Backends who implement directly will have no issue with that.
3. Drop deprecated methods from backends that support atomic APIs(when you can just implement atomic stuff).

So in the end will have that nothing was touched for backends that doesn't support atomic API and for backends that do, we'd have non-atomic methods be on the `dyn Window` trait impl directly, but not in the backend code itself.

dcz-self commented:
No failures now.

kchibisov commented on winit-core/src/window.rs:
```diff
@@ -1118,14 +1124,52 @@ pub trait Window: AsAny + Send + Sync + fmt::Debug {
+    /// - **iOS / Android / Web / Windows / X11 / macOS / Orbital:** Unsupported.
+    fn set_ime_state(&self, _state: Option<&ImeState>) {}
+
+    /// Returns the current IME state, if the set_ime_state API is supported.
+    fn get_ime_state(&self) -> Option<ImeState> {
+        None
+    }
```
Remove this part, no need to fix legacy API for Wayland. And we can fix it internally anyway.

dcz-self replied:
If I remove this, I have to remove the legacy API completely because it will stop being able to make set calls, see the other comments.

kchibisov commented on winit-core/src/window.rs:
```diff
@@ -1094,7 +1094,13 @@ pub trait Window: AsAny + Send + Sync + fmt::Debug {
-    fn set_ime_cursor_area(&self, position: Position, size: Size);
+    #[deprecated = "use set_ime_state instead"]
+    fn set_ime_cursor_area(&self, position: Position, size: Size) {
+        if let Some(state) = self.get_ime_state() {
```
Rewrite all of those without a `get_ime_state`, so just set it. That should follow how it was before.

kchibisov commented on examples/application.rs:
```diff
@@ -600,8 +605,8 @@ impl ApplicationHandlerExtMacOS for Application {
 struct WindowState {
-    /// IME input.
-    ime: bool,
+    /// State of the text input exposed to IME
+    ime_state: Option<ImeState>,
```
Can keep it as a bool.

dcz-self replied:
Same as above, the entire state must be cached.

kchibisov reviewed (changes requested):
I'd argue, that we shouldn't try to fix the legacy API we deprecate anyway.

dcz-self commented:
I think I see now where the confusion is coming from. The protocol allows a set of changes to be skipped and reused from the last commit.
That means every field is optional, and you fill it in depending on whether you support it and whether you want to update it. This makes it possible to bug out at runtime by sending a commit without required fields.

I decided to simplify it a little and instead fields are always filled in, except when optional. So winit takes care of persisting the properties for the protocol.
That's why the internals of the legacy API need to get_state: to get ImePurpose which is not optional.

I think this is better than the alternative because it allows winit to eliminate runtime errors, but if it's not a welcome solution, I will change it to the user being responsible for supplying the correct data manually.

kchibisov commented:
We just tell the user that they should submit the state entirely and manage it themselves, otherwise I think it's rather confusing.

kchibisov commented:
The API sets state, not adds state or joins state. Like logically it's much easier to reason about simpler concepts then keep in mind that you add state to winit rather than just set it entirely.

At the end of the day the entire state is set anyway, and winit can internally optimize communication with display server on platforms where it acts as addition.

dcz-self commented:
I agree that it's simpler to reason about APIs that do nothing but pass through calls, than add constraints. I just remember the case with the cursor_area where the Wayland protocol requires some value on first call, or otherwise following calls may get ignored. When the user gets a passthrough API, they need to read the protocol and know that detail, or their application will blow up on Wayland (I wrote the protocol and I forgot how it works, so I don't think others will be aware). When the user gets a (harder to reason about) API with constraints, they get pushed to do the thing that works.
Again, I'll send the simple API soon. I just wanted to explain my reasoning since you seem to be interested in discussing it.

kchibisov commented:
User will get event that IME is gone and so would have to re-issue on re-enable, so don't see an issue.

kchibisov commented:
It also means that once you re-enable it you adjust them, so if you have different caps based on the field you type in in your application, you will have better time understanding what you actually have now.

dcz-self commented:
> make enable not bool but rather set of capabilities telling client what they opt into

Sure, that makes sense.

There is a complication in that the protocol must receive *some* value in the first call. I assumed that if a client advertises a capability, then it will always use the data (and I still mostly think that's a correct assumption).

There are two ways to do that.

1. enable(CapabilitiesMap);

The first call to set_state() must set the data of the capabilities. If any capability is missing, a runtime check fails. The input method enable event is issued after first call.

2. enable(InitialState);

The initial state contains the supported capabilities. The enable event is set immediately. If set_state() is called with a capability that wasn't advertised, the user gets a runtime error.

What do you think?

kchibisov commented:
>There is a complication in that the protocol must receive some value in the first call. I assumed that if a client advertises a capability, then it will always use the data (and I still mostly think that's a correct assumption).

Even if it won't, winit can add some data just in case so it'll work. We just need a clear intention from the user, which is capability, since user might not have e.g. surrounding text around(I think my main problem is surrounding text, since I'd assume that if user doesn't pass it they don't have it yet).

kchibisov commented:
>The initial state contains the supported capabilities. The enable event is set immediately. If set_state() is called with a capability that wasn't advertised, the user gets a runtime error.

I think this makes a bit more sense, if user wants to enable something but doesn't enable feature while providing the data for it they should get an `assert`(of some sort, since it's a logical bug). On the other side, if they haven't set something e.g. `surrounding_text`, because they don't know it yet, we can just use e.g. default? Not sure how it's good for the end user though, could just assert both ways.

In general, my main problem right now is `surrounding_text` since for empty text field it's empty, but it also means that it's disabled then in protocol?

dcz-self commented:
> surrounding text, since I'd assume that if user doesn't pass it they don't have it yet).

The surrounding text can and should be an empty string if there's nothing around, so that's an easy one. The cursor area is more suspicious - I can't imagine a real use case where a text input sometimes has a cursor and sometimes doesn't, but maybe there is one.

With that explanation, should I implement option 2.?

kchibisov commented:
I was thinking by the time of setting to get a runtime error/assert if user supplied something that they haven't enabled, but IME won't get enabled, since it's more of an _assert_ I'd say, since it's a logical bug? So yeah, basically option 2, I guess.
And for legacy, we can just use placeholders, I guess and enable all 2 caps(surrounding text won't be part of it).

>The surrounding text can and should be an empty string if there's nothing around, so that's an easy one

Wouldn't it mean that it won't basically work once there will be something around according to the `text-input-v3` protocol? It says the following:

>The initial state for affected fields is empty, meaning that the text input does not support sending surrounding text. If the empty values get applied, subsequent attempts to change them may have no effect.

dcz-self commented:
Okay, I hit an issue again where it turns out I have to cache the entire state anyway.

The issue stems from the enabling model used by winit being different from the one used in Wayland.

With winit, it currently goes like this:
1. user enables IME on window
2. user receives IME_enabled event
3. user sets IME state

or
1. User enables IME on window
2. but there's no IME. the end.

Meanwhile, in Wayland:
1. application announces support for IME for all windows
2. compositor sigals IME availablity on a window with the "enter" event
3. user enables IME by sending initial state

or
1. application announces support for IME for all windows
2. but there's no IME. The compositor signals nothing, there's nowhere to send the event.

The problem is in the order of "enable" and "send state" requests. With the above conversation, we agreed to send initial state together with the "enable" request. But what if there's no IME yet?

There are two options:

1. winit is tracking the entire state, not just changes.

As soon as the IME becomes available, winit takes the state the user submitted and sends it as the initial state. This is the way it is currently in this MR.

2. winit is forwarding "IME enter" and "IME leave" events to the user.

If the user tries to enable the IME while no IME is available, then winit raises a runtime error. The job of handling IME availability falls on the user.
I don't like this option because it's not better, just different. It will break again when a platform comes that works in a different way.
I also don't think the user needs to know if an IME is available. IME updates look cheap to me and no harm is done if they are issued always, regardless of IME presence.

What do you think?

kchibisov commented:
I mean, winit should just cache the entire state and send it in that case when it can. Like there's nothing wrong with caching the entire thing in the backend, the problem is partial update of such state. So, just cache and send when ready.

kchibisov commented on winit-core/src/window.rs:
```diff
@@ -1530,6 +1631,93 @@ impl Default for ImePurpose {
     }
 }
 
+bitflags::bitflags! {
+    /// The user didn't set a property in the enabling call but tried to set it in an update.
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+    pub struct ImeUnsupportedCapability: u32 {
+        const PURPOSE = 0b1;
+        const CURSOR_AREA = 0b10;
+    }
+}
```
I mean that from the standpoint of the user, atomic means that you apply all your state at once, but what the current PR does is that you add state to winit state, and then it atomically sends it, which feels wrong, since how you remove things?

Like the problem is that it's additive and not _whole_.

dcz-self replied:
I see. The way I designed the Wayland interface, removing doesn't make sense because you must provide something for each capability. Once a capability is set, it's no longer optional until the input method is disallowed.
I did it that way because I can't imagine a situation where you set a capability and then not use it (ideally, all applications provide all info always, for all capabilities, but I realize that some people don't want to implement everything at once). I mean, the text field is not going to lose purpose (use normal purpose), and the caret isn't going to be not applicable (there's always a caret, unless we're not in a text field, but then disable the input method entirely).
Do you disagree with it? Should the general API allow properties to be optional even after advertising them in capabilities?

kchibisov replied:
I generally, I guess it's fine with partial updates to what already send, but then we don't need to cache at all if it's already stored on the server.

dcz-self replied:
What's the situation where it would not be set? Setting it to an empty string still counts as setting - the request goes out.

kchibisov replied:
Just to make it clear. The intent is to always send the state as it was enabled on initial request? So I always have to send the entire thing to the server?

dcz-self replied:
We don't need to cache it, but as I said before, it's simpler to reason about it when we cache it. We don't need to have two special modes: IME enabled (not storing), IME disabled (storing). Instead, we can just send out the complete state whenever the IME is available.
Plus, we still need to cache the capabilities all the time to check them at runtime, which is not much different from caching the entire state (it's storing a bool for every field, versus storing an Option for every field).
Do you think I should change that and instead implement two separate modes?

dcz-self replied:
> Just to make it clear. The intent is to always send the state as it was enabled on initial request? So I always have to send the entire thing to the server?

The intent of the Wayland protocol is that it's enough to send state updates.

The way I designed the winit support is that the complete state is sent to the server on every update. It's not intentional, it's just the simplest implementation in my opinion, if the library abstracts multiple platforms.
The user of the winit-core API still only sends updates, not the entire state.

kchibisov replied:
Hm, I think I'd prefer to send only part that user updates at the time, and for initial state based on capabilities either fill the blanks for what was not provided by the user(surrounding text). And internally not cache things other than caps.

kchibisov replied:
I can try to implement it myself so we don't back and forth on a design.

dcz-self replied:
Regarding filling the blanks, what do you mean by that?
When the enable API separates the capabilities and the state in two separate arguments, but they don't match?

dcz-self replied:
If you can implement this better, then go ahead, I'll review it. I just want this API to be merged because there's a bunch of properties that need to be added before it can be said this API is fully functional - and I think winit supporting the API would help me get things merged elsewhere.

dcz-self commented:
Thanks for taking care of this!

I like the ImeAction, it clearly shows that there's an order of submitting actions, and allows the user to explicitly understand what should be done next.

I think warnings should be there unconditionally - the way I see this design leading to annoying bugs is when the application developer mainly develops on Windows or whatever, and doesn't test very well on Wayland. Then a user encounters a bug where the IME popup is not appearing at all (capability not set), and the user is very confused. It only works if there are logs. Personally, I think it's better to force the developer to force to acknowledge the issue by forwarding the capabilities error all the way up.

I see you dropped the errors in a couple places, and I worry that the code is going to be a lot less readable once they are back in.

I think at least there should be an error when the user sets a capability but doesn't provide a value for it. Ideally, the error should be at compile time. Filling it in with random values isn't really making it work (cursor at top-left corner is not useful), it just allows the library user to forget/not understand things and be surprised at runtime, or worse, when the app reaches users.

I also think `ImeState` has the wrong name because it's not a state, it's a changeset.

But those are only suggestions. If it works the way you like, it works. I want to get it in and keep contributing.

dcz-self commented on winit-wayland/src/seat/text_input/mod.rs:
```diff
@@ -234,73 +221,56 @@ struct ContentType {
 pub struct ClientState {
-    content_type: Option<ContentType>,
+    /// Send enable.
+    enable: bool,
```
Is this used?

dcz-self commented:
> The point is not to submit those if winit doesn't have cap, I guess.

You mean when the backend doesn't support a certain capability? That that makes sense, but I still think logging is desired then, or else you can't debug it where it happens.
Or you could let the application developer decide how much they want to log and return a (not-error) struct saying which capabilities were rejected as unsupported.

Now I realize that perhaps making the cursor position optional was a mistake, everyone should support it.

kchibisov commented:
>Now I realize that perhaps making the cursor position optional was a mistake, everyone should support it.

If you don't want popup window visible as a user it could be a good option, don't see why not generally.

kchibisov commented:
I'd assume that if you have `warn!` logging in place, that should probably be enough for most cases, especially if we recommend backends to not perform operation when it was initially not enabled, which should make it more clear for the users where Wayland like system is not present.

Anyway, seems like you're in favor of proposed changes, thus I'll polish the PR and update the docs, etc.

kchibisov commented:
@dcz-self should be ready.

kchibisov commented on winit-wayland/src/seat/text_input/mod.rs:
```diff
@@ -189,11 +206,81 @@ pub struct TextInputDataInner {
+///
+/// Fields that are initially set to None are unsupported capabilities
+/// and trying to set them raises an error.
+#[derive(Debug, PartialEq, Clone, Default)]
```
I think that's intended? Like if Wayland wants non-sense to be sent then I think it's fine? Like there could be cases where you don't know what to send yet, but disabling the entire functionality just because of that sounds extensive. Again, it's a Wayland problem, not every other's problem.

dcz-self replied:
Sure. `Caps` is not a compile-time check, though. Not being able to accidentally and up with nonsense values because there's no `Default` is a compile-time property.

kchibisov replied:
Yeah, but it's still more clear then assumption based off fields, which you have to figure out. Like `Caps` is required as a type, so when you use it, you'll learn about fields and what they do. We can also add extra check to ensure that fields are actually set, but I'm not sure it's that needed?

kchibisov replied:
How it'll be different to `::new()`. Like the methods that accept IME changes both require `caps`, so you must have them in place. Like the point is that we don't have `Option` and rely on `Caps` for that, what is stored there is some default that will still make ime work if the user provided needed cap.

dcz-self replied:
You explained to me elsewhere why you're using a separate `Capabilities` struct, and I agree it makes sense in the public API, but I think in internal APIs `Option`s work better, just because with `Option`, the connection between the value and the availability of the value is very clear.

kchibisov commented on winit-wayland/src/window/state.rs:
```diff
@@ -983,51 +989,61 @@ impl WindowState {
+        request: ImeRequest,
+    ) -> Result<Option<bool>, ImeRequestError> {
+        let (capabilities, request_data, send_enable) = match request {
+            ImeRequest::Enable(capabilities, request_data) => {
```
it's a promise to winit, so the cross platform feel could be achieved and not doing X on Windows will lead to X not working on Wayland, and that could be consistently implemented by both backends and users.

Capabilities exist not to please input method, but rather give an abstraction to easily reason about without knowing semantics of Wayland and get the same behavior you'd get on Wayland on other backends. Like there could be values in the future, for which you'd have no idea what to provide initially, like e.g. surrounding text. Like the problem with initial approach is that figuring out what to set without caps is way harder than with caps, since you naturally discover them.

dcz-self replied:
> not doing X on Windows will lead to X not working on Wayland

I'm sorry, I'm not sure if I understand this. If the application doesn't do X, then sure, but then we don't have to worry that the application promises an X stream and then doesn't supply X.
If the Windows backend doesn't do X, then we have a backend-specific warning, so that will not lead to X not working on Wayland. But that doesn't tell me anything new, so I think I misunderstood what you meant.

> Like there could be values in the future, for which you'd have no idea what to provide initially

If there's a new value in the future supported by winit, then the old application will need a source code change to support it. It will not just get a random value inserted and sent to the IME, unless the IME author is an idiot and adds a new non-optional property, breaking compatibility for all their users.
Apart from that, I can't imagine a situation where an application developer wants to promise a property via capability and then not provide a value - except by mistake. Can you provide a case?

And I don't think surrounding text is one. Empty strings are an old concept in computing. The API already requires the user to send an empty surrounding string in this situation:
1. Text field with "text"
2. Focus and enable input method: set surrounding("text")
4. User deletes text
5. There's no request to unset surrounding except surrounding("")

so there's no value in special-casing the enable request and allowing uninitialized properties.

Maybe if I rephrase: I don't think capabilities are wrong. I just think this implementation will cause mistakes and pain to application developers.

Sorry for not making it clear: I think a better implementation will raise a runtime error if there's a capability/promise without an initial value.
An even better implementation will make it impossible at API level to provide a capability/promise without an initial value.

dcz-self commented on winit-core/src/window.rs:
````diff
@@ -1120,14 +1131,93 @@ pub trait Window: AsAny + Send + Sync + fmt::Debug {
+    /// window.request_ime_update(ImeRequest::Disable).expect("Disable cannot fail");
+    /// # }
+    /// ```
+    fn request_ime_update(&self, request: ImeRequest) -> Result<(), ImeRequestError>;
````
None of the three points above is described in `ImeRequest` docs, do you mean another `request`?

kchibisov replied:
The clearly were, except for `warn` part, I moved them a bit though.

kchibisov commented:
Could you look one more time?

dcz-self reviewed (commented):
You got it - it looks a lot better now.

dcz-self commented on winit-core/src/window.rs:
```diff
@@ -1530,6 +1630,185 @@ impl Default for ImePurpose {
+            return None;
+        }
+
+        if capabilities.contains(ImeCapabilities::PURPOSE) && request_data.cursor_area.is_none() {
```
`request_data.cursor_area.is_none()`

should be "purpose".

I think it would be easier to drop the `capabilities` argument because here you're checking the same thing twice.

kchibisov replied:
>Do you mean that you want the winit to advertise future properties that the user didn't opt into? The user won't be able to provide values for them anyway, so I must be misunderstading.

I meant here that future `ImeRequest::Update` won't work, if user won't pass `cursor_area`. But to figure that out and make backends more or less consistent, we need capabilities passed by the user rather than derived based off what they set initially, since in practice docs are not being read that often and you'll fix on a user report basis, while here we clearly communicate. 

I do understand that they are redundant though, it's just with something that generally lacks testing it's better to be more verbose.

dcz-self replied:
Got it, the redundancy is the point: it makes the user confirm the intention multiple times.

kchibisov commented:
Updated with tests ensuring that cap + provided data more or less match.
